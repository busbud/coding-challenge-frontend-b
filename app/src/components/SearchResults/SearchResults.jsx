import React, { useEffect, useState } from "react";
import { injectIntl, FormattedMessage } from "react-intl";
import CircularProgress from "@material-ui/core/CircularProgress";
import { getDepartures } from "../../services/getDepartures";
import SearchResult from "./SearchResult";

function SearchResults(props) {
  const [data, setData] = useState(null);
  const lang = props.intl.locale;

  useEffect(() => {
    let mounted = true;
    getDepartures(lang).then((result) => {
      if (mounted) {
        setData(result);
      }
    });
    return () => (mounted = false);
  }, [lang]);

  if (data) {
    if (data.departures && data.departures.length > 0) {
      let locations = new Map();
      data.locations.map(
        (location) => (locations[location.id] = location.name)
      );
      let cities = [];
      data.cities.map((city) => cities.push(city.name));
      let operatorLogos = new Map();
      let operatorNames = new Map();
      data.operators.forEach(function (operator) {
        operatorLogos[operator.id] = operator.logo_url.split("?")[0];
        operatorNames[operator.id] = operator.name;
      });
      return (
        <>
          {data.departures.map((departure) => (
            <SearchResult
              duration={departure.duration}
              originDepot={locations[departure.origin_location_id]}
              originCity={cities[0]}
              departTime={new Date(
                departure.departure_time
              ).toLocaleDateString(lang, {
                hour: "2-digit",
                minute: "2-digit",
              })}
              destinationDepot={locations[departure.destination_location_id]}
              destinationCity={cities[1]}
              arriveTime={new Date(
                departure.arrival_time
              ).toLocaleDateString(lang, {
                hour: "2-digit",
                minute: "2-digit",
              })}
              price={departure.prices.total}
              operatorLogo={operatorLogos[departure.operator_id]}
              operatorName={operatorNames[departure.operator_id]}
            />
          ))}
        </>
      );
    } else {
      return (
        <h2>
          <FormattedMessage id="no.results" />
        </h2>
      );
    }
  } else {
    return <CircularProgress />;
  }
}

export default injectIntl(SearchResults);
