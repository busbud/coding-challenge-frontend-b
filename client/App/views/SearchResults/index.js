import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Grid } from 'semantic-ui-react'
import SearchDetails from './components/SearchDetails'
import DeparturesList from './components/DeparturesList'
import { updateSearchResults } from '@/App/actions'
import './index.scss'

class SearchResults extends React.Component {
  constructor (props) {
    super(props)
    props.updateSearchResults()
  }

  render () {
    console.log('departures', this.props.departures)

    return (
      <Grid
        verticalAlign={'middle'}
        textAlign={'center'}
      >
        <Grid.Row>
          <Grid.Column mobile={15} tablet={9} largeScreen={6}>
            <div className="search-details">
              <SearchDetails
                origin={this.props.inputs.originCity.name}
                destination={this.props.inputs.destinationCity.name}
                date={this.props.inputs.date}
                adults={this.props.inputs.adults}
              />
            </div>
            <DeparturesList departures={this.props.departures}/>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    )
  }
}

const mapStateToProps = state => {
  return {
    inputs: state.inputs,
    departures: state.results.departures
  }
}

const mapDispatchToProps = dispatch => {
  return {
    updateSearchResults () {
      dispatch(updateSearchResults(...arguments))
    }
  }
}

const cityShape = PropTypes.shape({
  name: PropTypes.string.isRequired,
  geohash: PropTypes.string.isRequired
})

SearchResults.propTypes = {
  inputs: PropTypes.shape({
    originCity: cityShape,
    destinationCity: cityShape,
    date: PropTypes.instanceOf(Date).isRequired,
    adults: PropTypes.number
  }),
  departures: PropTypes.array,
  updateSearchResults: PropTypes.func
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SearchResults)
