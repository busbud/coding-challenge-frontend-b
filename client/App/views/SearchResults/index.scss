.search-results-wrap {
  height: 100%;
  padding-top: 20px;

  .grid {
    height: 100%;
  }
}

.search-details {
  margin-bottom: 20px;
}
