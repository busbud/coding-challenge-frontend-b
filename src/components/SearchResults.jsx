var React = require('react');
var Promise = require('bluebird');
var request = Promise.promisifyAll(require("request"));
var languages = require("../languages.js");

var GUEST_TOKEN = "GUEST_ZX3fNR26SB-m1MBsyz196g";
var API_URL = 'https://busbud-napi-prod.global.ssl.fastly.net';
var DEPARTURE_URL = "https://napi.busbud.com/x-departures/";

//Search Constants
var DEPARTURE_CITY = "New York";
var ARRIVAL_CITY = "Montreal";
var DATE = "2016-01-14";

//default city values for initial rendering of component
var departureCity = {
    geohash: "dr5reg",
    full_name: "New York, New York, United States"
};

var arrivalCity = {
    geohash: "f25dvk",
    full_name: "Montreal, Quebec, Canada"

};

//function to get a new token if the default token expires
//have not ran into this issue yet, so I am not sure the correct 
//error/response code to check for to implement this function
function getToken() {
    return request.getAsync("https://busbud-napi-prod.global.ssl.fastly.net/auth/guest")
        .then(function(response) {
            var guestToken = (JSON.parse(response.body)).token;
            return guestToken;
        });
}

function getCityInfo(city, language) {
    var options = {
        url: API_URL + "/search?q=" + city + "&lang=" + language,
        headers: {
            'x-busbud-token': GUEST_TOKEN
        }
    };
    return request.getAsync(options)
        .then(function(cityId) {
            return (JSON.parse(cityId.body))[0];
        });
}


var SearchResultsLayout = React.createClass({
            getInitialState: function() {
                return {
                };
            },
            componentDidMount: function() {
                var url = DEPARTURE_URL + this.props.departure.geohash + "/" + this.props.arrival.geohash + "/" + this.props.date + "/?lang=" + this.props.langAbbr;
                $.get(url, function(response){
                    console.log(response);
                    
                });
            },
            onClick: function(){
            },
            render: function() {
                    
            return (
                <section id="search-results">
                <div className="result-header">
                        <div className="row">
                            <div className="col-md-8 col-xs-10 col-lg-8 col-lg-push-2 col-xs-push-1 col-md-push-2">
                                <h4>Results For Your Trip </h4>
                                <h2> ____ to _____ </h2>
                                <h3>January 14, 2016</h3>
                            </div>
                        </div>
                    </div>
                    <div className="row">
                        <div className="results col-md-8 col-xs-12 col-lg-8 col-lg-push-2 col-md-push-2">
                            <div className="col-xs-7">
                                <div className="row">
                                    <h2>Departure</h2>
                                    <hr />
                                    <div className="results-column col-xs-4">
                                        <p>10:45pm</p>
                                    </div>
                                    <div className="results-column col-xs-8">
                                        <p>Montreal Bus Statation
                                            <br /> full address, H$4H5v
                                            <br />Montral Quebec
                                        </p>
                                    </div>
                                </div>
                                <div className="row">
                                    <h2>Arrival</h2>
                                    <hr />
                                    <div className="results-column col-xs-4">
                                        <p>1:45am</p>
                                    </div>
                                    <div className="results-column col-xs-8">
                                        <p>NEW YORK CITY
                                            <br /> full address, H$4H5v
                                            <br />Montral Quebec
                                        </p>
                                    </div>
                                </div>
                            </div>
                            <div className="results-column col-xs-5">
                                <div className="results-column col-xs-12">
                                    <h2>Outbound Price</h2>
                                    <hr />
                                    <p className="price">
                                        $205.00<span className="currency"> CAD</span>
                                    </p>
                
                                </div>
                                <div className="results-column col-xs-12">
                
                                    <hr />
                                    <p>
                                        Image Logo
                                        <br />
                                        <br /> Economy
                                    </p>
                                </div>
                
                            </div>
                        </div>
                    </div>
            </section>
            );
        }
});

module.exports = SearchResultsLayout;