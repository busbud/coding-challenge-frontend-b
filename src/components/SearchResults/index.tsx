import * as React from 'react'

export default class SearchResults extends React.Component<any, any> {
  constructor (props: any) {
    super(props)

  }

  render () {
    return(
      <h1>Search results</h1>
    )
  }
}
