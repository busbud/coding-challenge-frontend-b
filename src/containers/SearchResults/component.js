import React from 'react';
import { array, object } from 'prop-types';
import { compareAsc } from 'date-fns';
import { map, compose, sortWith, path, ascend } from 'ramda';
import Grid from 'material-ui/Grid';
import Trip from '../../components/Trip';

const propTypes = {
  classes: object.isRequired,
  cities: array,
  departures: array,
};

const makeTip = ({ id, ...props }) => <Trip key={id} {...props} />;

const departureTimeASC = (a, b) => compareAsc(new Date(a.departure_time), new Date(b.departure_time));
const priceASC = ascend(path(['prices', 'total']));

const mapTrip = compose(map(makeTip), sortWith([departureTimeASC, priceASC]));

const Component = ({ classes, departures }) => (
  <Grid container className={classes.root}>
    {mapTrip(departures)}
  </Grid>
);

Component.propTypes = propTypes;

export default Component;
