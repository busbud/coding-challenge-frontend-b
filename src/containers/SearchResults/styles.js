export default {
  root: {
    marginTop: '6rem',
    marginBottom: '3rem',
  },
};
