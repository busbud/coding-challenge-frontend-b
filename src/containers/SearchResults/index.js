import { connect } from 'react-redux';
import { compose } from 'recompose';
import { createStructuredSelector } from 'reselect';
import { withStyles } from 'material-ui/styles';
import styles from './styles';
import { makeSelectStructuredDepartures } from '../App/selectors';
import component from './component';

export const mapStateToProps = createStructuredSelector({
  departures: makeSelectStructuredDepartures(),
});

export const enhance = compose(withStyles(styles), connect(mapStateToProps, null));

export default enhance(component);
