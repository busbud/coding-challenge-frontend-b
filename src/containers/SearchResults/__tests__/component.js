import React from 'react';
import { shallow, render } from 'enzyme';
import SearchResults from '../component';

jest.mock('../../../components/Trip', () => ({ id, ...props }) => <div key={id} {...props} />);

const props = {
  classes: {
    root: 'fakeRootClass',
  },
  departures: [
    {
      id: '1',
      departure_time: '2018-03-10T08:30:00',
      prices: {
        total: 100,
      },
    },
    {
      id: '2',
      departure_time: '2018-03-10T07:30:00',
      prices: {
        total: 100,
      },
    },
    {
      id: '3',
      departure_time: '2018-03-10T07:30:00',
      prices: {
        total: 90,
      },
    },
    {
      id: '4',
      departure_time: '2018-03-10T06:30:00',
      prices: {
        total: 100,
      },
    },
  ],
};

describe('component | SearchResults | component', () => {
  describe('render', () => {
    it('should render without throwing and error', () => {
      // given
      const wrapper = shallow(<SearchResults {...props} />);

      // then
      expect(wrapper).toBeDefined();
    });

    it('should match snapshot', () => {
      // given
      const wrapper = shallow(<SearchResults {...props} />);

      // then
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Grid', () => {
    it('should have proper className props', () => {
      // given
      const wrapper = shallow(<SearchResults {...props} />);
      const { className } = wrapper.props();

      // then
      expect(className).toBe(props.classes.root);
    });
  });
  describe('Trips', () => {
    it('should render correct number of trip', () => {
      // given
      const wrapper = render(<SearchResults {...props} />);
      const { length } = wrapper.children();

      // then
      expect(length).toBe(4);
    });
    it('should render trips in correct order', () => {
      // given
      const wrapper = shallow(<SearchResults {...props} />);
      expect(wrapper.childAt(0).key()).toBe(props.departures[3].id);
      expect(wrapper.childAt(1).key()).toBe(props.departures[2].id);
      expect(wrapper.childAt(2).key()).toBe(props.departures[1].id);
      expect(wrapper.childAt(3).key()).toBe(props.departures[0].id);
    });
  });
});
