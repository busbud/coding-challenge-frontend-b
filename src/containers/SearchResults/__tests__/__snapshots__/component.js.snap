// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component | SearchResults | component render should match snapshot 1`] = `
<WithStyles(GridWrapper)
  className="fakeRootClass"
  container={true}
>
  <Component
    departure_time="2018-03-10T06:30:00"
    key="4"
    prices={
      Object {
        "total": 100,
      }
    }
  />
  <Component
    departure_time="2018-03-10T07:30:00"
    key="3"
    prices={
      Object {
        "total": 90,
      }
    }
  />
  <Component
    departure_time="2018-03-10T07:30:00"
    key="2"
    prices={
      Object {
        "total": 100,
      }
    }
  />
  <Component
    departure_time="2018-03-10T08:30:00"
    key="1"
    prices={
      Object {
        "total": 100,
      }
    }
  />
</WithStyles(GridWrapper)>
`;
